File: python/backtracking/combination_sum.py
"""
Problem: Combination Sum
Given an array of distinct integers candidates and a target, return all unique combinations where the chosen numbers sum to target. Same number may be chosen unlimited times.

Approach:
1. Backtracking with pruning.
"""
from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        # TODO: implement
        raise NotImplementedError

if __name__ == "__main__":
    print(Solution().combinationSum([2,3,6,7], 7))  # expect [[2,2,3],[7]]

File: cpp/backtracking/combination_sum.cpp
/*
Problem: Combination Sum
Approach: Backtracking with pruning.
*/
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        // TODO: implement
        return {}; // placeholder
    }
};

int main() {
    vector<int> c = {2,3,6,7};
    Solution s; auto res = s.combinationSum(c, 7);
    for (auto &v: res) { cout << '['; for (size_t i=0;i<v.size();++i){cout<<v[i]<<(i+1<v.size()?",":"");} cout << "] "; }
    cout << endl; // expect [2,2,3] [7]
}
